[AMDGPU] Lazily emit waitcnts on function entry

jayfoad opened the pull request:
Instead of emitting s_waitcnt 0 on entry to a function, just remember
that the counters are in an unknown state and let the pass emit the
waitcnts naturally before the first use of any register that might need
it.

Delay the s_waitcnts could have a small performance improvement, and in
some cases allows them to be combined with other waits which can have a
small code size improvement.

jayfoad committed:
[AMDGPU] Lazily emit waitcnts on function entry

Instead of emitting s_waitcnt 0 on entry to a function, just remember
that the counters are in an unknown state and let the pass emit the
waitcnts naturally before the first use of any register that might need
it.

Delaying the s_waitcnts could have a small performance improvement, and
in some cases allows them to be combined with other waits which can have
a small code size improvement.

jayfoad commented:
This is a WIP. Lots of tests still need to be updated and some of the diffs look dubious.

jayfoad commented on llvm/test/CodeGen/AMDGPU/extract-subvector-16bit.ll:
```diff
@@ -33,14 +32,21 @@ define <4 x i16> @vec_8xi16_extract_4xi16(ptr addrspace(1) %p0, ptr addrspace(1)
+; SI-NEXT:    s_mov_b64 s[4:5], 0
+; SI-NEXT:    s_andn2_b64 vcc, exec, s[4:5]
+; SI-NEXT:    s_waitcnt lgkmcnt(0)
+; SI-NEXT:    s_mov_b64 vcc, vcc
```
This is very poor code. Maybe the placement of the waitcnt interfered with some peephole optimization that was suppose to clean it up?

jayfoad replied:
Right, `SIInsertWaitcnts::insertWaitcntInBlock` generates `s_waitcnt lgkmcnt(0)` and `s_mov_b64 vcc, vcc` to work around a bug on GFX6 where vccz could be clobbered by completion of an SMEM load. Then `SIPreEmitPeephole::optimizeVccBranch` fails to look through the mov.

This only affects GFX6 so I don't think it's worth tring to fix it.

jayfoad commented on llvm/test/CodeGen/AMDGPU/llvm.amdgcn.set.inactive.chain.arg.ll:
```diff
@@ -11,20 +11,18 @@
 ; GFX11-NEXT:    s_not_b32 exec_lo, exec_lo
 ; GFX11-NEXT:    global_store_b32 v[8:9], v0, off
-; GFX11-NEXT:    s_nop 0
-; GFX11-NEXT:    s_sendmsg sendmsg(MSG_DEALLOC_VGPRS)
```
This needs investigating.

jayfoad replied:
This change is good. This patch naturally fixes the same bug that #72245 fixes.

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/call-outgoing-stack-args.ll:
```diff
@@ -223,9 +223,9 @@ define amdgpu_kernel void @kernel_caller_byval() {
 ; MUBUF-LABEL: func_caller_stack:
 ; MUBUF:       ; %bb.0:
-; MUBUF-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
 ; MUBUF-NEXT:    s_mov_b32 s4, s33
 ; MUBUF-NEXT:    s_mov_b32 s33, s32
```
This is assuming that SP/FP are never written by a memory instruction, which is probably an OK assumption but we should probably document it

jayfoad replied:
I don't want to change the ABI in this patch so I have "fixed" this by checking for reserved SGPRs and marking them as possibly depending on any of the counters.

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/clamp-fmed3-const-combine.ll:
```diff
@@ -4,8 +4,8 @@
 ; GFX10:       ; %bb.0:
-; GFX10-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
 ; GFX10-NEXT:    v_mul_f32_e64 v0, v0, 2.0 clamp
+; GFX10-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
```
We can't guarantee that v0 isn't being written by a pending load

jayfoad replied:
Good point. This was a major thinko. I am now marking live-ins to the entry block as possibly depending on any of the counters.

jayfoad committed:
[AMDGPU] Simplify WaitcntBrackets::getRegInterval with getPhysRegBaseClass

This means that getRegInterval no longer depends on the MCInstrDesc, so
it could be simplified further to take just a MachineOperand or just a
physical register. NFCI.

jayfoad committed:
Add a getRegInterval overload that just takes a Register

jayfoad commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -292,6 +295,56 @@ class WaitcntBrackets {
+
+    // Reserved SGPRs (e.g. stack pointer or scratch descriptor) may depend on
+    // any counter.
+    // FIXME: Why are these not live-in to the function and/or the entry BB?
```
Is there a good reason for this?

arsenm replied:
SP is always manually managed without memory operations, so I don't think there's any useful way it can depend on a counter. If the caller didn't have FP/BP and was using it as a general SGPR, it's maybe possible they were used with a memory operation we need to complete before initializing them in the prolog

jayfoad commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -292,6 +295,56 @@ class WaitcntBrackets {
+
+    // Live-in registers may depend on any counter.
+    const MachineBasicBlock &EntryMBB = MF.front();
+    for (auto [Reg, Mask] : EntryMBB.liveins()) {
```
I tried reimplementing this part using `MRI.liveins` instead of `EntryMBB.liveins` but apparently VGPR function arguments are not included in the former. Is that expected?

rovka commented on llvm/test/CodeGen/AMDGPU/GlobalISel/assert-align.ll:
```diff
@@ -6,10 +6,12 @@ declare hidden ptr addrspace(1) @ext(ptr addrspace(1))
+; CHECK-NEXT:    s_waitcnt lgkmcnt(0)
 ; CHECK-NEXT:    s_mov_b32 s16, s33
 ; CHECK-NEXT:    s_mov_b32 s33, s32
+; CHECK-NEXT:    s_waitcnt expcnt(0)
```
Why are we waiting for expcnt here? (as opposed to after changing EXEC)

jayfoad replied:
Quoting from `SIInsertWaitcnts`:
```
  // Export & GDS instructions do not read the EXEC mask until after the export
  // is granted (which can occur well after the instruction is issued).
  // The shader program must flush all EXP operations on the export-count
  // before overwriting the EXEC mask.
```

rovka replied:
Right, thanks! Found it in the SPG too, it just wasn't in the table where I was originally looking.

rovka commented on llvm/test/CodeGen/AMDGPU/GlobalISel/assert-align.ll:
```diff
@@ -44,7 +46,6 @@ entry:
 define ptr addrspace(1) @tail_call_assert_align() {
 ; CHECK-LABEL: tail_call_assert_align:
 ; CHECK:       ; %bb.0: ; %entry
-; CHECK-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
```
Why do these drop completely here, but in the next test (`br-constant-invalid-sgpr-copy.ll`) they just move to the end of the function? Is it because this function returns something, so it's the caller's job to wait before reading the results?

jayfoad replied:
This function ends with a tail call, not a return. The ABI says that counters are undefined on function entry but must be 0 on return.

rovka commented on llvm/test/CodeGen/AMDGPU/GlobalISel/br-constant-invalid-sgpr-copy.ll:
```diff
@@ -8,24 +8,24 @@
 ; WAVE64-NEXT:    s_cmp_lg_u32 s4, 0
 ; WAVE64-NEXT:    s_cbranch_scc1 .LBB0_1
 ; WAVE64-NEXT:  ; %bb.2: ; %.exit5
+; WAVE64-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
```
Why exactly do we need to wait here? We only wrote some SGPRs that we're not returning

jayfoad replied:
As above, counters are undefined on function entry but must be 0 on return.

rovka reviewed (commented):
Should `callWaitsOnFunctionEntry` still return true all the time?

jayfoad commented on llvm/test/CodeGen/AMDGPU/GlobalISel/add.v2i16.ll:
```diff
@@ -171,15 +171,15 @@ define <2 x i16> @v_add_v2i16_neg_inline_imm_splat(<2 x i16> %a) {
 ; GFX9-LABEL: v_add_v2i16_neg_inline_imm_splat:
 ; GFX9:       ; %bb.0:
-; GFX9-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
 ; GFX9-NEXT:    v_mov_b32_e32 v1, 0xffc0ffc0
```
Yes I think you're right. I guess that's another fundamental problem with this patch :(

jayfoad commented on llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:
```diff
@@ -292,6 +295,56 @@ class WaitcntBrackets {
+    // FIXME: Why are these not live-in to the function and/or the entry BB?
+    for (unsigned I = 0, E = ST->getMaxNumSGPRs(MF); I != E; ++I) {
+      MCRegister Reg = AMDGPU::SGPR_32RegClass.getRegister(I);
+      if (MRI.getReservedRegs()[Reg])
```
It's a trivial accessor defined in the header so I don't think there's much point.
